### ISSUE TITLE
[Docs] A walkthrough manifolds/stereographic functions 💥

### ISSUE THREAD

**Issue by lancejpollard** [geoopt/geoopt#204]
Looking at [geoopt/manifolds/stereographic/math.py](https://github.com/geoopt/geoopt/blob/27a1e6fd825e2de923b8493bb8baf514f3f089ea/geoopt/manifolds/stereographic/math.py) and trying to learn about hyperbolic geometry as I go, wondering if you could add some commentary to the functions. If anyone is in the mood to teach, I am all ears :). I can prompt with some questions, hopefully that will help clarify what is confusing or hard to understand for a newcomer.

1. `artanh`: [Area](https://en.wikipedia.org/wiki/Inverse_hyperbolic_functions) functions, or angle functions. Don't quite understand what this means yet, but I can probably figure it out.
1. [`artan_k`](https://github.com/geoopt/geoopt/blob/27a1e6fd825e2de923b8493bb8baf514f3f089ea/geoopt/manifolds/stereographic/math.py#L215)
1. [`tan_k`](https://github.com/geoopt/geoopt/blob/27a1e6fd825e2de923b8493bb8baf514f3f089ea/geoopt/manifolds/stereographic/math.py#L191): What is this?
1. [`tan_k_zero_taylor`](https://github.com/geoopt/geoopt/blob/27a1e6fd825e2de923b8493bb8baf514f3f089ea/geoopt/manifolds/stereographic/math.py#L52): What is this?
1. [`artan_k_zero_taylor`](https://github.com/geoopt/geoopt/blob/27a1e6fd825e2de923b8493bb8baf514f3f089ea/geoopt/manifolds/stereographic/math.py#L90)
1. [`abs_zero_grad`](https://github.com/geoopt/geoopt/blob/27a1e6fd825e2de923b8493bb8baf514f3f089ea/geoopt/manifolds/stereographic/math.py#L46)
1. `project`: Safe projection on the manifold for numerical stability. **What does that mean exactly? Is it used for rendering?**
1. `lambda_x`: Compute the conformal factor $\lambda^\kappa_x$ for a point on the ball. **What is this used for?** https://en.wikipedia.org/wiki/Conformal_map
1. `inner`: Compute inner product for two vectors on the tangent space w.r.t Riemannian metric on the Poincare ball. What is the use of this? It makes sense what it does but not sure how it gets applied.
1. `norm`: Compute vector norm on the tangent space w.r.t Riemannian metric on the Poincare ball. **What is the use of this?**
1. `mobius_add`: Compute the Möbius gyrovector addition. **Why do we do this?**
1. `mobius_coadd`: Compute the Möbius gyrovector coaddition. **What is the difference between add and coadd?** What is this used for?
1. `mobius_sub`: Compute the Möbius gyrovector subtraction.
1. `mobius_cosub`: Compute the Möbius gyrovector cosubtraction.
1. `mobius_scalar_mul`: Compute the Möbius scalar multiplication.
1. `mobius_matvec`: Compute the generalization of matrix-vector multiplication in gyrovector spaces. **What is this used for**?
1. `mobius_pointwise_mul`: Compute the generalization for point-wise multiplication in gyrovector spaces. **What is this use case?**
1. `mobius_fn_apply`: Compute the generalization of function application in gyrovector spaces. **What is this used for?**
1. `mobius_fn_apply_chain`: Compute the generalization of sequential function application in gyrovector spaces.
1. `mobiusify`: Wrap a function such that is works in gyrovector spaces.
1. `gyration`: Compute the gyration of $u$ by $[a,b]$. **How does this function get used / taken advantage of?**
1. `dist`: Compute the geodesic distance between $x$ and $y$ on the manifold. **This one I understand.** A geodesic is a path along a manifold. So compute the distance along that path.
1. `dist0`: Compute geodesic distance to the manifold's origin. **Distance to origin**.
1. `dist2plane`: Geodesic distance from $x$ to a hyperplane $H_{a, b}$. **How is this different than just `dist`?**
1. `geodesic`: Compute the point on the path connecting $x$ and $y$ at time $x$. Say again, this is a point? How is ti different than the `dist` function?
1. `geodesic_unit`: Compute the point on the "unit speed geodesic". **What does that mean / how is it used?** https://en.wikipedia.org/wiki/Geodesic
1. `expmap`: Compute the exponential map of $u$ at $x$. **What is the "exponential map" used for?** https://en.wikipedia.org/wiki/Exponential_map_(Riemannian_geometry)
1. `expmap0`: Compute the exponential map of $u$ at the origin $0$.
1. `logmap`: Compute the logarithmic map of $y$ at $x$. **How does this compare to the exponential map?** What is it used for?
1. `logmap0`: Compute the logarithmic map of $y$ at the origin $0$.
1. `parallel_transport`: Compute the parallel transport of $v$ from $x$ to $y$. **What is "[parallel transport](https://en.wikipedia.org/wiki/Parallel_transport)" and what is it used for?**
1. `parallel_transport0`: Compute the parallel transport of $v$ from the origin $0$ to $y$.
1. `parallel_transport0back`: Perform parallel transport to the zero point. (Optimization when last point is zero?)
1. `egrad2rgrad`: Convert the Euclidean gradient to the Riemannian gradient. **What is a gradient in this context?** And how does the gradient get used?
1. `sproj`: Stereographic Projection from hyperboloid or sphere. **So this maps the hyperboloid onto the sphere for like rendering?**
1. `inv_sproj`: Inverse of Stereographic Projection to hyperboloid or sphere. **What does inverse mean in this context?** From X to hyperboloid?
1. `antipode`: Compute the antipode of a point $x_1,...,x_n$ for $\kappa > 0$. What are [antipodal points](https://mphitchman.com/geometry/section6-1.html). What are they used for?
1. `weighted_midpoint`: Compute weighted Möbius gyromidpoint. **What is this used for?**

This is basically all the functions in that main.py, and wondering how to apply them / what they mean from a practical perspective.

I am trying to build a cellular automaton in the hyperbolic space (visualization), and would like to use your [Universal Model for Hyperbolic, Euclidean and Spherical Geometries](https://andbloch.github.io/K-Stereographic-Model), but need to learn these foundational elements first, without hopefully having to spend 5 years relearning linear algebra and then onto the foundations of hyperbolic geometry in a 1000 page textbook. If I could understand what these functions mean at a high practical level, then it would be easier to see how to apply them, and learn from them.

Thank you so much for your time and project!

**Comment by ferrine** [comment]
Hi, I'm happy you find the project useful. I'm currently not actively develop the project, but still maintain it. I can help to teach you how to contribute to docs and improve the documentation. Are you interested in opening a PR?

**Comment by lancejpollard** [comment]
Yeah I can totally add some docs once I understand what to add. Maybe you can just walk me through some stuff and I'll convert it into docs?

**Comment by ferrine** [comment]
Writing docs is nothing but adding docstrings to functions in [RST](https://docs.fileformat.com/programming/rst/) format.

https://github.com/geoopt/geoopt/blob/master/geoopt/manifolds/stereographic/math.py

In case you need plots, you can use the following example (I have a [dedicated place](https://github.com/geoopt/geoopt/tree/master/docs/plots/extended/stereographic) to keep plotting mess):

https://github.com/geoopt/geoopt/blob/master/geoopt/manifolds/stereographic/math.py#L462

As for hyperbolic geometry overview, I believe [this](https://arxiv.org/abs/1911.05076) paper is what you need. You can read it to understand better what K stands for and how other trigonometric functions are involved.

Since the document was more than just writing docstrings for functions It is recomposed into an more educational oriented format

https://github.com/geoopt/geoopt/blob/master/docs/extended/stereographic.rst

The document is again RST and can reference arbitrary functions like here

https://github.com/geoopt/geoopt/blob/master/docs/extended/stereographic.rst?plain=1#L127

Everything is rearranged or omitted for the readability purpose (It took some time to write the first version of the document and I never came back to it)
